### ISSUE TITLE
Docker Setup Questions

### ISSUE THREAD

**Issue by emmakat** [1337-server/automatic-ripping-machine#65]
### Description
I'm struggling with documentation of the docker setup and was hoping to share these in case anyone might be able to help. 

Thank you for all your support and work on this amazing project! All of the early issues I faced indeed had to do with permissions issues, as well as the machine not having the arm user created 1st, so I did that as a fresh start.

**build image** 

1. the arm user does not have access to modify the /opt directory and using sudo creates another issue where the cloned files are owned by root. To get around this and ease of use, I changed the permissions recursively for opt. This is not ideal and may have introduced a vulnerability but this machine is not used for anything else so I take the risk. 

2. did the git clone as instructed, and chmod +x all the files and chmod -R 777 on arm/docs. I then did the chmod +x arm/scripts/docker_build.sh and then ran the docker_build.sh and this all works fine.

_Here's where I get confused._ 
"Once the image completes successfully, you can then run/add the container with the docker run command .... Remember to allow -p 8080 inside the container to something outside." - I don't understand exactly what to type here - should I look to the other steps to "run a docker manually" or some subset of that?

3. **setup/docker-setup.sh** errors out because the arm folder is already created and is not empty. I looked at the file, should I just run the part at the bottom in order to proceed?
install setup/docker-arm.rules /etc/udev/rules.d/docker-arm.rules
udevadm control --reload

4. **Setup drives step** - I know how to do this and it works fine however - the dockerfile seems to contain these steps within it. So, should it be done separately and if so, should that be removed from the dockerfile or are they both necessary?

**5. prebult docker image**
I'm confused about the difference at the end of the steps for the docker run commands in the first example vs the commands in the pre-built instructions. Why are they different and what is the result of the different steps?
_build/manual run:_
--cap-add SYS_ADMIN \
    --security-opt apparmor:unconfined \
    --restart "always" \
    --name "arm-rippers" \
    arm-combined:latest

 _pre-built_
 --name "arm-rippers" \
   1337server/automatic-ripping-machine:latest

**6. Changing where docker puts files**
I've attempted multiple times to instruct arm to point to a 2nd storage drive for the paths, and even though I see the dockerfile assumes/becomes the arm user, and thus should have permissions to create files in the storage drive, it simply doesn't work and ripping doesn't happen. What confuses me is that the -v command is used for binding AND for volume, so I should in theory be able to change the mapping to point to the other drive. But since the container just doesn't rip anything when I attempt to use the drive like so:  
   -v "/media/arm/storage:/home/arm" \
   -v "/media/arm/storage/media/music:/home/arm/Music" \
   -v "/media/arm/storage/config:/home/arm/config" \
   -v "/media/arm/storage/logs:/home/arm/logs" \
   -v "/media/arm/storage/media:/home/arm/media" \

### Environment
[OS Distribution and version (run "cat /etc/lsb-release")]
[ARM Release Version or if cloning from git branch/commit (run "git branch" and "git log -1" to get this)]

### Log file
[Run the rip in DEBUG and drag and drop the log file onto this comment box]

[>>>>> DRAG AND DROP LOG FILE HERE <<<<<]

**Comment by emmakat** [comment]
OK! Here's what I did to fix all of my issues, and using the build image option.
1. fresh install of Ubuntu 20.10, create the arm user, install lsscsi
2. set permissions for /opt so that the git clone can work without sudo
3. modify docker-entrypoint.sh

- _to add another variable for the storage drive_

`export STORAGE="/media/arm/storage/"`

- _change the needed/expected dirs to point to the new storage variable_

`# setup needed/expected dirs if not found
SUBDIRS="config media media/completed media/raw media/movies encode logs db Music .MakeMKV"
for dir in $SUBDIRS ; do
  thisDir="${STORAGE}/${dir}"`
I had to use sudo chmod -R 777 against the storage directory. I might try adding that step in the docker-entrypoint.sh like `sudo chmod -R 777 ${STORAGE}`

4. modify docker-setup.sh 
_comment out the git clone step and the mkdir step since they were done previously, and fail otherwise_ then run `install setup/docker-arm.rules /etc/udev/rules.d/docker-arm.rules`

`#git clone -b docker --depth=1 https://github.com/1337-server/automatic-ripping-machine.git "$SRCDIR"`
`#mkdir -p "$DATADIR"`

6. modified the Dockerfile to add additional drives since I have a total of 7 drives

Finally, I built the image, and it's happily ripping from every drive. Later, I will re-image and re-run everything fresh to ensure I didn't miss any steps/mods, and will update this issue.

**Comment by 1337-server** [comment]
5.
```--cap-add SYS_ADMIN```
 ``` --security-opt apparmor:unconfined ``` are both from deekue's pull request/repo they are what's recommended however i couldn't get ARM to rip with just these, but i didn't want to fully remove them.  Personally i could only ever get it running with the `--privileged` flag added, but my setup is not standard so that may be why i was experiencing issues. 

I thought it better to include them so that if someone with more experience with docker and its build process tries to use it they could correct the wiki. 




> Here's where I get confused.
> "Once the image completes successfully, you can then run/add the container with the docker run command .... 
> Remember to allow -p 8080 inside the container to something outside." - I don't understand exactly what to type here - > should I look to the other steps to "run a docker manually" or some subset of that?

This is a very good point, i used portainer when first trying out docker. That was just to remind myself that i needed to map port 8080, as i would remember to map the folders but forget to map the port. More than once i was left scratching my head as to what i managed to break in the docker, only to find i had forgot to map the port.
But this probably only adds more confusion. I'll remove this.

**Comment by emmakat** [comment]
Thank you so much for your response!! I don't know that any of this will be useful but here is the readme for my forked copy https://github.com/emmakat/automatic-ripping-machine/blob/jessica/README.md

**Comment by 1337-server** [comment]
This is fantastic and very clear! With your permission i would like to add it to the wiki. My idea is that i create a new page for the docker build guide (using your readme) as the docker build and have a page for just pulling the image. It should make things much easier to read/follow.

Just let me know if you're happy for it to be added with/without credits and i will add it.

**Comment by emmakat** [comment]
you can use any/all of it! No attribution needed. I am guessing you'd want to make a page that copies just using the portions of my readme that are applicable? I think the storage variable instructions will be useful to others but the instructions should list them as options vs assuming it. 

Also I'm worried about the setup udev section being busted right now. I'd like to fix it but I don't really know what's wrong. Before you proceed, I'd like to re-run the entire thing to verify everything early tomorrow morning - since I also broke my .abcde.conf file really badly somehow and now it's decided to stop ripping things.

**Comment by emmakat** [comment]
Here is a first draft for you to take a look at, and see if it would work for your docker build guide :) 
[docker_build_guide.txt](https://github.com/1337-server/automatic-ripping-machine/files/6291330/docker_build_guide.txt)

**Comment by 1337-server** [comment]
That looks great! If you let me know when you think your done with it all add it to the wiki. 

Also your abcde file looks **much** easier to read, the only downside is that with comments removed unless you know what its doing, it might be more confusing.

Installing eyeD3, metaflac, and AtomicParsley into the docker should be very simple as modifying the Dockerfile to include these in the apt/pip install.

**Comment by emmakat** [comment]
Ok after much thrashing, and staying up too late, I'm attaching a fresh copy that seeks to clarify the docker instructions. This leaves out any of the suggestions for changing the docker entrypoint for pointing to a 2nd drive because after I ran through everything, I found I couldn't repeat the setup reliably. 
[docker_build_guide.txt](https://github.com/1337-server/automatic-ripping-machine/files/6300709/docker_build_guide.txt)

I sent in a PR for the additions to the dockerfile, and it only needs eyed3 and atomicparsley, the metaflac wasn't needed apparently.
